Add Acmeda integration

atmurray opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This pull request adds an integration for Rollease Acmeda Automate Blinds. The integration was modelled off the Hue integration. It is implemented by using the PyPI module "aiopulse" to talk over the local network using a proprietary protocol. When the user adds it, the integration attempts to discover any hubs on the local network. For the selected hub, any blinds connected to the hub are added as a cover and a sensor (for battery state of charge).

This resolves the following feature request:
https://community.home-assistant.io/t/rollease-acmeda-automate-pulse-hub-integration

## Docs PR:
https://github.com/home-assistant/home-assistant.io/pull/13331

## Type of change
- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
This is my first crack at developing for home-assistant and have tried to follow all the developer guidelines and checklists. Please let me know if I've gotten anything wrong.

- Feature request: https://community.home-assistant.io/t/rollease-acmeda-automate-pulse-hub-integration
- Uses PiPI module aiopulse: https://pypi.org/project/aiopulse/ hosted in github: https://github.com/atmurray/aiopulse
- Only tested for version 1 hubs (a version 2 hub exists but is not known to work for this).
- Thread to discuss integration: https://community.home-assistant.io/t/integration-for-rollease-acmeda-automate-pulse-hub-and-blinds
- Both integration and module use asyncio
- I have an outstanding question that may impact this integration: https://community.home-assistant.io/t/creating-persistent-async-tasks

## Checklist
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

atmurray committed:
First cut of Rollease Acmeda Pulse Hub integration.

atmurray committed:
Merge remote-tracking branch 'upstream/dev' into acmeda

atmurray committed:
Acmeda integration improvements:
 - Moved common code into a base entity
 - Battery level sensor added
 - Localisation now working

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray committed:
Added requirement for aiopulse now that it has been uploaded to PyPI.

atmurray committed:
Exclude acmeda integration from coverage check as it relies on a hub being present.

homeassistant commented:
Hi @atmurray, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#33384).

Once you do that we will be able to review and accept this pull request.

Thanks!

frenck commented:
Hi there @atmurray!

Thanks for opening up a PR, on a first quick glance to see if everything is complete:

- Tests for the configuration flow are missing. Configuration flows require tests with a full 100% coverage on the flow itself.
- Documentation is missing.
- Builds are not passing, make sure it passes.

Thanks! 👍

atmurray committed:
Fix Travis CI build issues.

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray commented:
Hi @frenck,
> * Tests for the configuration flow are missing. Configuration flows require tests with a full 100% coverage on the flow itself.
No problem, will get onto this.
> * Documentation is missing.
...and this
> * Builds are not passing, make sure it passes.
I think I've fixed some of the build failures however it looks like the plex component is causing some tests to fail. Is this a known issue in the dev branch at the moment?
https://travis-ci.com/github/atmurray/home-assistant/jobs/307556709

bdraco commented on homeassistant/components/acmeda/cover.py:
```diff
@@ -0,0 +1,230 @@
+    cur_covers = {}
+    # cur_groups = {}
+
+    # Acmeda updates all covers via a single API call.
```
This seems like a good use case for `DataUpdateCoordinator`

https://developers.home-assistant.io/docs/integration_fetching_data/#coordinated-single-api-poll-for-data-for-all-entities

atmurray replied:
Not sure I understand. The api is fully asynchronous and isn’t polled. There is a callback to notify hass that the entity has been updated.

bdraco replied:
At first glace it looked like this was updating data from the hub.  I see now that you are sending a batch of updates.

bdraco commented on homeassistant/components/acmeda/const.py:
```diff
@@ -0,0 +1,10 @@
+DOMAIN = "acmeda"
+
+DEFAULT_PORT = 12416
+DEFAULT_SCAN_INTERVAL = timedelta(seconds=60)
```
```suggestion
SCAN_INTERVAL = timedelta(seconds=60)
```

bdraco replied:
Actually looks like its unused

bdraco commented on homeassistant/components/acmeda/const.py:
```diff
@@ -0,0 +1,10 @@
+LOGGER = logging.getLogger(__package__)
+DOMAIN = "acmeda"
+
+DEFAULT_PORT = 12416
```
```suggestion
DEFAULT_PORT = 12416
```
Looks unused

atmurray replied:
Done.

bdraco commented on homeassistant/components/acmeda/cover.py:
```diff
@@ -0,0 +1,230 @@
+            )
+        )
+
+        # tasks.append(
```
Please remove all commented code

bdraco commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,111 @@
+    host = entry.data["host"]
+    config = hass.data[DATA_CONFIGS].get(host)
+
+    if config is None:
```
pass or pass ?

bdraco commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,239 @@
+        # tasks.append(
+        #    async_update_items(
+        #        hass,
+        #        config_entry,
```
Please remove commented code

bdraco commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,239 @@
+    await update_hub()
+
+
+async def async_update_items(
```
This looks similar to the code in cover.py.  Can you break this out so its not repeated?

atmurray replied:
Done, refactored into `helpers.py`.

bdraco commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,239 @@
+            attr.update(super_attr)
+
+        attr[ATTR_BATTERY_LEVEL] = self.roller.battery
+        attr[ATTR_BATTERY_CHARGING] = False
```
```suggestion
        attr[ATTR_BATTERY_CHARGING] = False
```
If this doesn't support charging, please remove.

bdraco commented on homeassistant/components/acmeda/cover.py:
```diff
@@ -0,0 +1,230 @@
+    async def request_update(is_group, object_id):
+        """Request an update.
+
+        We will only make 1 request to the server for updating at a time. If a
```
Can this be moved into the `aiopulse` package so Home Assistant doesn't have to know about this complexity?

atmurray replied:
Good suggestion, have enhanced aiopulse so it will notify the integration of any changes to the hub configuration. Now using `helpers/dispatcher.py` to notify the entities that they need to check for new instances.

bdraco commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,46 @@
+            "identifiers": {(DOMAIN, self.device_id)},
+            "name": self.roller.name,
+            "manufacturer": "Rollease Acmeda",
+            # "model": self.roller.productname or self.roller.modelid,
```
Please remove commented out code

atmurray replied:
Done

bdraco commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,45 @@
+        entity = current[item_id]
+        removed_items.append(item_id)
+        await entity.async_remove()
+        ent_registry = await get_ent_reg(hass)
```
await entity.async_remove() in `helpers/entity.py` is doing this under the hood.

It looks like much of the below can be removed.

```

    async def async_remove(self) -> None:
        """Remove entity from Home Assistant."""
        assert self.hass is not None
        await self.async_internal_will_remove_from_hass()
        await self.async_will_remove_from_hass()

        if self._on_remove is not None:
            while self._on_remove:
                self._on_remove.pop()()

        self.hass.states.async_remove(self.entity_id, context=self._context)


```

atmurray replied:
I couldn't work out how to have the entity completely removed from the user interface without this code. The code was copied from the hue integation in `components/hue/helpers/py`.

bdraco commented on homeassistant/components/acmeda/hub.py:
```diff
@@ -0,0 +1,110 @@
+        # register service
+        results = await asyncio.gather(
+            self.hass.config_entries.async_forward_entry_unload(
+                self.config_entry, "cover"
```
Do you need to unload "sensor" here as well?

atmurray replied:
Resolved.

bdraco commented on homeassistant/components/acmeda/hub.py:
```diff
@@ -0,0 +1,110 @@
+
+        self.api = hub
+
+        hass.async_create_task(
```
It might be helpful to keep track of this in a constant so you can use it for unload as well

```
    for platform in PLATFORMS:
        hass.async_create_task(
            hass.config_entries.async_forward_entry_setup(config_entry, platform)
        )
```

atmurray committed:
Remove unused constants.

atmurray committed:
Remove unused group logic from cover.py

atmurray committed:
Removed commented code from base.py

atmurray committed:
Remove sensors (battery entities) on removal of hub.

atmurray committed:
Remove unused groups from sensor.py

atmurray committed:
Acmeda device and entity update made fully asynchronous using subscriptions to remove need for config polling.

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray committed:
Updated aiopulse version dependency.
Removed non-functional battery charging indication.

atmurray committed:
Rationalised common code to update entities into helpers.py

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray commented:
Documentation added:
https://github.com/home-assistant/home-assistant.io/compare/current...atmurray:patch-1

atmurray committed:
Fix linting issue.

bdraco commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,74 @@
+    @property
+    def name(self):
+        """Return the name of roller."""
+        return super().name + " Battery"
```
```suggestion
        return f"{super().name} Battery"
```

The Home Assistant standard is to use f-strings.

atmurray replied:
Resolved.

bdraco commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,101 @@
+        host = hub_conf[CONF_HOST]
+
+        # Store config in hass.data so the config entry can find it
+        hass.data[DATA_CONFIGS][host] = hub_conf
```
Please import the data into the config flow instead

atmurray replied:
Done.

bdraco commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,101 @@
+        if host in configured_hosts:
+            continue
+
+        # No existing config entry found, trigger link config flow. Because we're
```
Stale comment?

atmurray replied:
Removed.

bdraco commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,101 @@
+    if not await hub.async_setup():
+        return False
+
+    hass.data[DOMAIN][host] = hub
```
```suggestion
    hass.data[DOMAIN][entry.entry_id] = hub
```

Ideally everything is indexed by the entry_id as we know its going to be unique and the user can't accidentally setup two with the same name

atmurray replied:
Done.

bdraco commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        # This is for backwards compatibility.
+        return await self.async_step_init(user_input)
+
+    async def async_step_init(self, user_input=None):
```
```suggestion
    async def async_step_user(self, user_input=None):
```
Please use step_user and delete the above

https://developers.home-assistant.io/docs/config_entries_config_flow_handler/

atmurray replied:
Done.

bdraco commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            if self.unique_id is None:
+                await self.async_set_unique_id(hub.id, raise_on_progress=False)
+
+            return self.async_create_entry(title=hub.host, data={"host": hub.host},)
```
```suggestion
            return self.async_create_entry(title=hub.host, data={"host": hub.host})
```

atmurray replied:
Removed.

bdraco commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+            self.hub = self.discovered_hubs[user_input["id"]]
+            await self.async_set_unique_id(self.hub.id, raise_on_progress=False)
+            # We pass user input to link so it will attempt to link right away
+            return await self.async_step_link({})
```
```suggestion
            return await self.async_step_link()
```

Doesn't look used

atmurray replied:
Removed.

bdraco commented on homeassistant/components/acmeda/hub.py:
```diff
@@ -0,0 +1,127 @@
+        try:
+            # Create a Hub object and verify connection.
+            try:
+                # self.task = hass.async_create_task(hub.run())
```
```suggestion
```

atmurray replied:
Removed.

bdraco commented on homeassistant/components/acmeda/hub.py:
```diff
@@ -0,0 +1,127 @@
+
+        self.api = hub
+
+        self.tasks.append(
```
The PluseHub really shouldn't know about setting up config entries if it can be avoided.  Please do this outside of `class PulseHub`

bdraco replied:
```
    for component in PLATFORMS:
        hass.async_create_task(
            hass.config_entries.async_forward_entry_setup(config_entry, component)
        )
```

atmurray replied:
Moved to __init__.py

atmurray committed:
Correct additional CI pylint errors.

bdraco commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,58 @@
+    def notify_update(self):
+        """Tell HA that the device has been updated."""
+        LOGGER.debug("Device update notification received: %s", self.name)
+        super().schedule_update_ha_state()
```
If this is a callback from async, add `@callback` and you can write the state directly.
```suggestion
        self.async_write_ha_state()
```

bdraco replied:
`from homeassistant.core import callback`

atmurray replied:
Done.

bdraco commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,122 @@
+        discovered bridges. Triggered for any bridge that does not have a
+        config entry yet (based on host).
+
+        This flow is also triggered by `async_step_discovery`.
```
Stale comment?

atmurray replied:
Updated.

atmurray committed:
Index config_entries by entry_id.
Move entity loading and unloading to __init__.py
Add entry_id to dispatcher signal
Removed now unused polling code hub
Added config_flow unit tests

atmurray committed:
Tweak to integration config_entry title.

atmurray committed:
Bumped aiopulse module to 0.3.2.
Reduced verbosity of aiopulse module.

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray committed:
Changed to using direct write of device state.
Removed old style async_step_init config_flow step.

bdraco commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,69 @@
+        """Return the name of roller."""
+        return f"{super().name} Battery"
+
+    @property
```
Please remove `battery_level` and `device_state_attributes` as its already in the state.   We only need it once

atmurray replied:
Done.

atmurray committed:
Remove superfluous battery_level and device_state_attributes from battery entity.

bdraco commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,81 @@
+            LOGGER.debug("Removing %s %s", entity_class.__name__, unique_id)
+            removed_items.append(element)
+
+    for entity in removed_items:
```
@MartinHjelmare I'm not 100% sure on this code. Can you take a look?

atmurray replied:
I've incorporated his feedback. Thanks.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,81 @@
+    hass, entity_class, config_entry, current, async_add_entities
+):
+    """Add any new entities, remove any old ones."""
+    if config_entry.entry_id not in hass.data[DOMAIN]:
```
This can't happen unless we make a bug.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,81 @@
+        if unique_id not in current:
+            LOGGER.debug("New %s %s", entity_class.__name__, unique_id)
+            new_item = entity_class(hass, roller)
+            current[unique_id] = new_item
```
Don't store the entity in a shared container. Entities are not allowed to be accessed outside their platforms and we should also not access an entity from outside itself directly so there shouldn't be a need to store them.

It's ok to store a reference to the entity eg the entity_id.

atmurray replied:
Done, now only storing unique_id of the entity.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,81 @@
+    for entity in removed_items:
+        # Device is removed from Pulse, so we remove it from Home Assistant
+
+        await entity.async_remove()
```
Signal the entity to remove itself via our dispatch helper instead.

atmurray replied:
Done, moved removal of entity and de-registration from device and entity registry to the base class. This is signalled using the dispatch helper as suggested.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,81 @@
+            )
+
+
+def create_config_flow(hass, host):
```
This doesn't seem used.

atmurray replied:
Removed

MartinHjelmare commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,118 @@
+            continue
+
+        # No existing config entry found, trigger link config flow. Because we're
+        # inside the setup of this component we'll have to use hass.async_add_job
```
Stale comment. Just remove this comment. It's standard for all integrations with import to the config flow.

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,118 @@
+    for hub_conf in hubs:
+        host = hub_conf[CONF_HOST]
+
+        if host in configured_hubs:
```
We don't need this check. We do it in the config flow too.

atmurray replied:
Removed

MartinHjelmare commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,60 @@
+class AcmedaBase(entity.Entity):
+    """Base representation of an Acmeda roller."""
+
+    def __init__(self, hass, roller: aiopulse.Roller):
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

atmurray replied:
Removed

MartinHjelmare commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,60 @@
+    @property
+    def device_id(self):
+        """Return the ID of this roller."""
+        return self.unique_id
```
Use `self.roller.id` instead. We don't want to mix unique_id and device id. The unique_id is the unique id of the entity.

atmurray replied:
Changed.

MartinHjelmare commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,60 @@
+    def device_info(self):
+        """Return the device info."""
+        return {
+            "identifiers": {(DOMAIN, self.roller.id)},
```
Maybe use `self.device_id`?

atmurray replied:
Changed.

MartinHjelmare commented on homeassistant/components/acmeda/base.py:
```diff
@@ -0,0 +1,60 @@
+        """Entity has been added to hass."""
+        self.roller.callback_subscribe(self.notify_update)
+
+    async def async_reset(self):
```
Rename this to `async_will_remove_from_hass`.

atmurray replied:
Changed.

MartinHjelmare commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,119 @@
+
+            if hub.id:
+                title = f"{hub.id} ({hub.host})"
+            else:
```
If there's not hub id we can't set unique_id for the config entry?

atmurray replied:
The challenge I found was that if the hub was configured in the yaml then we won't know the hub.id until it connects for the first time. Should I mandate that that id is provided in the yaml and reject the connection if the specified ID doesn't match what the device reports?

MartinHjelmare replied:
We should check the hub id after connection also for the import flow and abort if that id is already set up.

atmurray replied:
So should it try and connect during the import flow to obtain the hub id?
If so, what should it do if it can't currently connect?

The current approach is asynchronous such that the config flow can complete even before the first successful connection is made. This means that on startup devices that aren't responding are still shown in the integrations and can be removed by the user.

MartinHjelmare replied:
We should always check that we can connect before creating a config entry. If we can't connect during the import flow, log error and abort.

atmurray replied:
I implemented doing a connect before creating a config entry but found that this stalled the starting of home assistant. Noting that YAML configuration is deprecated I opted to remove the YAML config import support.

MartinHjelmare commented on homeassistant/components/acmeda/cover.py:
```diff
@@ -0,0 +1,60 @@
+                hass, AcmedaCover, config_entry, current, async_add_entities
+            )
+
+    async_dispatcher_connect(
```
We should unsubscribe the dispatcher in `async_unload_entry` in the base integration module. Store the return value here in hass.data, and use the returned function to unsubscribe.

atmurray replied:
Added a cleanup_callback list to the PulseHub class.
The returned functions are called on async_reset of the PulseHub object.

MartinHjelmare commented on homeassistant/components/acmeda/manifest.json:
```diff
@@ -0,0 +1,14 @@
+  "config_flow": true,
+  "documentation": "https://www.home-assistant.io/integrations/acmeda",
+  "requirements": ["aiopulse==0.3.2"],
+  "ssdp": [],
```
Empty items should be removed.

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/sensor.py:
```diff
@@ -0,0 +1,48 @@
+    @property
+    def icon(self):
+        """Return the icon to use in the frontend, if any."""
+        return icon_for_battery_level(battery_level=self.roller.battery)
```
Remove icon property when we set device class for the entity.

atmurray replied:
Removed

MartinHjelmare commented on homeassistant/components/acmeda/strings.json:
```diff
@@ -0,0 +1,31 @@
+      "unknown": "Unknown error occurred",
+      "cannot_connect": "Unable to connect to the hub",
+      "already_configured": "Hub is already configured",
+      "already_in_progress": "Config flow for hub is already in progress.",
```
Where is this string used?

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/strings.json:
```diff
@@ -0,0 +1,31 @@
+      "cannot_connect": "Unable to connect to the hub",
+      "already_configured": "Hub is already configured",
+      "already_in_progress": "Config flow for hub is already in progress.",
+      "not_pulse_bridge": "Not a Pulse hub"
```
As above.

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/strings.json:
```diff
@@ -0,0 +1,31 @@
+      "no_hubs": "No Pulse hubs discovered",
+      "all_configured": "All Pulse hubs are already configured",
+      "unknown": "Unknown error occurred",
+      "cannot_connect": "Unable to connect to the hub",
```
Cannot connect isn't used.

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/strings.json:
```diff
@@ -0,0 +1,31 @@
+      "discover_timeout": "Unable to discover Pulse hubs",
+      "no_hubs": "No Pulse hubs discovered",
+      "all_configured": "All Pulse hubs are already configured",
+      "unknown": "Unknown error occurred",
```
Unknown string isn't used.

atmurray replied:
Removed.

MartinHjelmare commented on homeassistant/components/acmeda/strings.json:
```diff
@@ -0,0 +1,31 @@
+      }
+    },
+    "error": {
+      "register_failed": "Failed to register, please try again",
```
Register failed string isn't used.

atmurray replied:
Removed.

atmurray committed:
Removal of unused strings.
Removal of unused create_config_flow helper.
Removal of stale comment.

atmurray committed:
Remove use of shared container to track existing enities.
Moved removal and deregistration of entities to base class through use of dispatch helper.

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray committed:
Fixed strings.json

atmurray committed:
Fix incorrect use of remove instead of pop on dict.

atmurray committed:
Add support for tilting covers, bump aiopulse version number.

atmurray committed:
Bump aiopulse version to v0.3.4.
Fixed bug in cover supported_features.

bdraco commented on .coveragerc:
```diff
@@ -10,6 +10,7 @@ omit =
     # omit pieces of code that rely on external devices being present
     homeassistant/components/acer_projector/switch.py
     homeassistant/components/actiontec/device_tracker.py
+    homeassistant/components/acmeda/*
```
We still want to track `config_flow.py` so please leave that in and add entries for the files that are missing tests.

atmurray replied:
Done.

atmurray committed:
Bumped aiopulse version to 0.4.0
Update acmeda .coveragerc exclusions

atmurray committed:
Merge branch 'dev' of https://github.com/home-assistant/home-assistant into acmeda

atmurray committed:
Removed already configured hub check from __init__.py async_setup_entry
Removed passing in hass reference to base entity class
Renamed entity async_reset to async_will_remove_from_hass
Changed device_info and properties
Migrated to CoveEntity from CoverDevice
Added dispatched_connect cleanup on hub removal
Removed unused entries from manifest
Removed override of battery icon
Renamed translations folder

atmurray committed:
Reversed unintended change to .coveragerc

atmurray committed:
Fixed config flow for multi-hub discovery.

MartinHjelmare commented on homeassistant/components/acmeda/__init__.py:
```diff
@@ -0,0 +1,105 @@
+    if conf is None:
+        conf = {}
+
+    hass.data[DOMAIN] = {}
```
```suggestion
```

atmurray replied:
Removed

MartinHjelmare commented on homeassistant/components/acmeda/hub.py:
```diff
@@ -0,0 +1,88 @@
+
+            for unique_id in list(self.current_rollers):
+                if unique_id not in self.api.rollers:
+                    LOGGER.info("Notifying remove of %s", unique_id)
```
Debug max.

atmurray replied:
Changed to debug.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,37 @@
+async def add_entities(hass, entity_class, config_entry, current, async_add_entities):
+    """Add any new entities."""
+    hub = hass.data[DOMAIN][config_entry.entry_id]
+    LOGGER.info("Looking for new %s on: %s", entity_class.__name__, hub.host)
```
```suggestion
    LOGGER.debug("Looking for new %s on: %s", entity_class.__name__, hub.host)
```

atmurray replied:
Changed to debug.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,37 @@
+    new_items = []
+    for unique_id, roller in api.items():
+        if unique_id not in current:
+            LOGGER.info("New %s %s", entity_class.__name__, unique_id)
```
```suggestion
            LOGGER.debug("New %s %s", entity_class.__name__, unique_id)
```

atmurray replied:
Changed to debug.

MartinHjelmare commented on homeassistant/components/acmeda/helpers.py:
```diff
@@ -0,0 +1,37 @@
+from .const import DOMAIN, LOGGER
+
+
+async def add_entities(hass, entity_class, config_entry, current, async_add_entities):
```
This can be a callback since we don't await inside.
```suggestion
@callback
def async_add_acmeda_entities(hass, entity_class, config_entry, current, async_add_entities):
```

atmurray replied:
Changed to callback, similarly changed calling functions in sensor and cover.

MartinHjelmare commented on homeassistant/components/acmeda/cover.py:
```diff
@@ -0,0 +1,122 @@
+    current = set()
+
+    async def async_update():
+        async with update_lock:
```
I don't think we need the lock. Only one of the jobs can run at any time in the event loop. And we'll move to use a callback.

atmurray replied:
Removed locks.

atmurray committed:
Acmeda enhancements as requested by MartinHjelmare

MartinHjelmare commented:
Looks good. One unresolved comment above remains and some build checks are failing.

atmurray committed:
Force import to connect to hub to retrieve id prior to creating entry

atmurray committed:
Remove YAML configuration support.

MartinHjelmare commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,80 @@
+            ),
+        )
+
+    async def async_step_create(self, user_input=None):
```
This doesn't have to be a step method with optional user input. It can be a helper method that accepts the hub as parameter.

atmurray replied:
I've tidied up the config flow and removed an unnecessary step, I think this is in line with what you've suggested.

MartinHjelmare commented on tests/components/acmeda/test_config_flow.py:
```diff
@@ -0,0 +1,121 @@
+
+
+async def async_generator(items):
+    """Async ytields items provided in a list."""
```
yields

atmurray replied:
Ftixed.

MartinHjelmare commented on tests/components/acmeda/test_config_flow.py:
```diff
@@ -0,0 +1,121 @@
+
+
+async def test_already_configured(hass, mock_hub_discover):
+    """Test that flow aborts when all habs are configured."""
```
hubs

atmurray replied:
Fixed.

atmurray committed:
Tidied up config_flow and tests:
 - removed unnecessary steps
 - fixed typos

atmurray committed:
Removed storage of hub in config_flow.

atmurray commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+            ),
+        )
+
+    async def async_create(self, hub=None):
```
Removed.

atmurray commented on homeassistant/components/acmeda/config_flow.py:
```diff
@@ -0,0 +1,77 @@
+
+    def __init__(self):
+        """Initialize the config flow."""
+        self.hub: Optional[aiopulse.Hub] = None
```
Removed.

MartinHjelmare reviewed (approved):
Looks good!
